### DISCUSSION TITLE
TheXTech v1.3.5.2: More than just a bugfix update

### DISCUSSION THREAD

**Discussion by Wohlstand** [Wohlstand/TheXTech#214] [General]
This update should've been a bugfix update, however, it introduces some set of new nice features you would like!

**Note:** To get the full power of TheXTech while making new projects, the special [Moondust Devkit configuration package](https://github.com/Wohlstand/TheXTech-SDK) is suggested (the latest laboratory Moondust Devkit required until the next stable devkit version 0.4.3.1 release).

## Significant changes
- Added an option to enable the background game controller input handling
- Portal warps have been allowed to be used for level entrances and level exits (Instant warps kept unsupported to avoid possible compatibility issues)
- Added the cannon pipes that can shoot the player with a given speed
- Added the opt-in 2x scale down optimization for all textures to reduce the memory usage
- License was changed into GPLv3 (see details: https://github.com/Wohlstand/TheXTech/issues/162)
- Enabled the work on the Android TV
- Added mask fallback support for custom graphics mask compatibility for old episodes
- Added an ability to change the look of the launcher screen on Android via gameinfo.ini (Change the logo, change the background, and set up the simple animation)
- Added wireless game controllers battery status displaying when enabled at the settings
- Added device battery status displaying at the right-top corner when enabled at the settings
- Fixed an inability to save the game when game assets were stored at SD Card on an Android device
- Added the Yoshi mode support for specified songs at the music.ini: mute selected track number without Yoshi, enable it back when Yoshi is presented
- Added a fixed behavior for rail platforms and grinder (NPC-60, NPC-62, NPC-64, NPC-66, NPC-104, NPC-179). It can be enabled using new-added custom values

## Changes for 1.3.5.2
- Fixed the ARM64 building
- Updated the log writer
- Disabled the "Boss Dead" event hook to stop the episode speed-run timer by default, added the compat.ini options to enable and configure it
- Fixed the unexpected Swooper flap sound played after the Swooper has gone far away
- Fixed the vanilla bug: The Swooper has avoided being started to fly when inactive
- Added the compatibility level command line option separated from the speed-run modes (However, the speed-run mode will still override the compativility level)
- Added an option to enable the background game controller input handling
- Portal warps have been allowed to be used for level entrances and level exits (Instant warps kept unsupported to avoid possible compatibility issues)
- Added the cannon pipes that can shoot the player with a given speed
- Added the workaround for custom background2-42 that has non-multiple 4 height (there are several custom sprites with the 3455 invalid height, they must be 3466)
- Fixed an incorrect speed-run timer render at multiplayer mode
- The controller state print has been replaced with a controller drawing to indicate the pressure state of buttons
- Added the opt-in 2x scale down optimization for all textures to reduce the memory usage
- Fixed the confusion when using the modern section auto-scroll settings when an auto-scroll fix was disabled
- License was changed into GPLv3 (see details: https://github.com/Wohlstand/TheXTech/issues/162)
- Enabled the work on the Android TV
- Fixed the game start error after granting the storage permission on Android 9 and lower
- Added an ability to change the look of the launcher screen on Android via gameinfo.ini (Change the logo, change the background, and set up the simple animation)
- Added mask fallback support for custom graphics mask compatibility for old episodes
- Fixed vanilla bug: Attempt to enter the locked warp or take the star with star-locked warp the game will crash if a level has no BGO at all
- Make turn blocks (block-90) work as bricks when SMBX19 and lower level file version is open
- Added a workaround for the yellow platform (NPC-60) at "The Invasion 1" episode where it won't move at level42
- Added Billy Gun (NPC-22) behavior compatibility with SMBX 1.2 and 1.1.x and older file formats
- Added wireless game controllers battery status displaying when enabled at the settings
- Added device battery status displaying at the right-top corner when enabled at the settings
- Fixed an inability to save the game when game assets were stored at SD Card on an Android device
- Added support for gamecontrollerdb.txt at the assets root and at the user directory
- Fixed vanilla bug: World map music won't resume after going through the warp that leads to no music points or the music point with the same song
- Fixed the improper custom music change when going through the warp on the world map
- Fixed vanilla bug: Offscreen NPC can now respawn during the stopwatch freeze
- Added Thwomp (NPC-37) behavior compatibility with SMBX 1.0.x
- Added the Yoshi mode support for specified songs at the music.ini: mute selected track number without Yoshi, enable it back when Yoshi is presented
- Fixed a bug when a last world map music is not remembered when sound is disabled
- Fixed the wrong restart place when running the lasanha of levels and "restart level after death" option is enabled
- Fixed the NPC logic distortion caused by early gfx update
- Added an ability to customize an outro behaviour using gameinfo.ini
- Battle mode access can be disabled via gameinfo.ini
- Two-player game mode access can be disabled via gameinfo.ini
- Added a fixed behavior for rail platforms and grinder (NPC-60, NPC-62, NPC-64, NPC-66, NPC-104, NPC-179). It can be enabled using new-added custom values
- Fixed the collision bug when an attempt of the player with the shoe to jump up at the moving NPC being pushed into the wall, causes the player to go through the wall
- The settings/thextech.ini config file will be created on the first launch if not exist

## Known issues
* Affected by [DrPepper Problem](https://wohlsoft.ru/pgewiki/DrPepper_Problem): The strong lag on slow devices when the level has any moving layers that move horizontally (by X-axis), mainly because the blocks search indexing gets completely disabled when the layer gets moved by X-axis. Todo: implement the QuadTree for all level objects and make sure they properly getting updated and found. (Issue #33, got been resolved at the `devel` branch by @ds-sloth)
* Can't type any cheats on a mobile device without a USB-OTG keyboard yeh. Some levels (for example, the "Superb Demo 128" at "Prelude To The Stupid") do require the use of cheats for being passed, otherwise, these levels becoming unbeatable without a USB-OTG keyboard plugged. Todo: implement the calling of the software keyboard to allow running cheats on mobile devices. (Issue #61)

## Showcases
![Scr_2021-06-11_03-08-49](https://user-images.githubusercontent.com/6751442/137928893-48d9e872-cc39-4ccc-a49a-ab989ba2313c.gif)

![Scr_2021-06-11_03-16-04](https://user-images.githubusercontent.com/6751442/137929013-4f020268-b913-4e5a-ad62-ddac78e064dd.gif)

![Scr_2021-09-09_23-47-01](https://user-images.githubusercontent.com/6751442/137929529-1ff61af4-3d2c-45bc-9de4-3fd00a14b396.gif)

![Scr_2021-10-18_02-14-18](https://user-images.githubusercontent.com/6751442/137929759-9e38c70d-853f-474b-8a95-1df77fbdfd5c.gif)

<hr /><em>This discussion was created from the release <a href='https://github.com/Wohlstand/TheXTech/releases/tag/v1.3.5.2'>TheXTech v1.3.5.2: More than just a bugfix update</a>.</em>

**Comment by LoveBodhi** [comment]
Seems my asset also require to be updated (for V1.1).

Patch for TheXTech SDK will be planned, too.
